test/hpke_test.c

```diff
@@ -0,0 +1,1728 @@
+    enclen = 200;
+    suite.kem_id = OSSL_HPKE_KEM_ID_X25519; /* back to default */
+    suite.aead_id = 0x1234; /* bad aead */
+    if (!TEST_false(OSSL_HPKE_get_grease_value(NULL, NULL, &suite, &suite2,
+                                               enc, &enclen, ct, ctlen)))
+        return 0;
+    enclen = 200;
+    suite.aead_id = def_suite.aead_id; /* good aead */
+    suite.kdf_id = 0x3451; /* bad kdf */
+    if (!TEST_false(OSSL_HPKE_get_grease_value(NULL, NULL, &suite, &suite2,
+                                               enc, &enclen, ct, ctlen)))
+        return 0;
+    enclen = 200;
+    suite.kdf_id = def_suite.kdf_id; /* good kdf */
+    suite.kem_id = 0x4517; /* bad kem */
+    if (!TEST_false(OSSL_HPKE_get_grease_value(NULL, NULL, &suite, &suite2,
+                                               enc, &enclen, ct, ctlen)))
+        return 0;
+    return 1;
+}
+
+/*
+ * @brief generate a key pair from an initial string and check public
+ * @param kem_id the KEM to use (RFC9180 code point)
+ * @ikm is the initial key material buffer
+ * @ikmlen is the length of ikm
+ * @pub is the public key buffer
+ * @publen is the length of the public key
+ * @return 1 for good, other otherwise
+ *
+ * This calls OSSL_HPKE_keygen specifying only the IKM, then
+ * compares the key pair values with the already-known values
+ * that were input.
+ */
+static int test_hpke_one_ikm_gen(uint16_t kem_id,
+                                 unsigned char *ikm, size_t ikmlen,
+                                 unsigned char *pub, size_t publen)
+{
+    OSSL_HPKE_SUITE hpke_suite = OSSL_HPKE_SUITE_DEFAULT;
+    unsigned char lpub[OSSL_HPKE_TSTSIZE];
+    size_t lpublen = OSSL_HPKE_TSTSIZE;
+    EVP_PKEY *sk = NULL;
+
+    hpke_suite.kem_id = kem_id;
+    if (!TEST_true(OSSL_HPKE_keygen(hpke_suite, lpub, &lpublen, &sk,
+                                    ikm, ikmlen, testctx, NULL)))
+        return 0;
+    if (!TEST_ptr(sk))
+        return 0;
+    EVP_PKEY_free(sk);
+    if (!TEST_mem_eq(pub, publen, lpub, lpublen))
+        return 0;
+    return 1;
+}
+
+static int test_hpke_ikms(void)
+{
+    int res = 1;
+
+    res = test_hpke_one_ikm_gen(0x20,
+                                ikm25519, sizeof(ikm25519),
+                                pub25519, sizeof(pub25519));
+    if (res != 1)
+        return res;
+
+    res = test_hpke_one_ikm_gen(0x12,
+                                ikmp521, sizeof(ikmp521),
+                                pubp521, sizeof(pubp521));
+    if (res != 1)
+        return res;
+
+    res = test_hpke_one_ikm_gen(0x10,
+                                ikmp256, sizeof(ikmp256),
+                                pubp256, sizeof(pubp256));
+    if (res != 1)
+        return res;
+
+    res = test_hpke_one_ikm_gen(0x10,
+                                ikmiter, sizeof(ikmiter),
+                                pubiter, sizeof(pubiter));
+    if (res != 1)
+        return res;
+
+    return res;
+}
+int setup_tests(void)
```

Comment:
In order to test your libctx is working...

Add globals at the top..
static OSSL_LIB_CTX *libctx = NULL;
static OSSL_PROVIDER *nullprov = NULL;


Add this to setup_tests()
    if (!test_get_libctx(&libctx, &nullprov, NULL, NULL, "default"))
        return 0;
        
Add 
void cleanup_tests(void)
{
    OSSL_LIB_CTX_free(libctx);
    OSSL_PROVIDER_unload(nullprov);
}

And then use this libctx... What this does is if you pass NULL as the libctx it will use a NULL provider.. Which will fail to fetch anything.. Using the supplied libctx it fetches from the default provider.

Reply:
libctx = testctx in your case.. (since you already have a global).

Reply:
Note also that if you specify a propq for one of your tests then it will hit that branch in coverage also,,
"provider=default" should work..

Reply:
So I think that's done now. I didn't add any command line arguments to really try out a non-default libctx though. I guess that could be added later?

Reply:
Not really needed..